mtt-management/src/mtt_types.rs

```diff
@@ -0,0 +1,446 @@
+// to by the L2 entry determines the type.
+// The 12-bit index into the 4K page is computed using the 14-bits [25:12] of the physical address, and then shifting right by 2
+#[allow(dead_code)]
+fn get_l1_entry(phys_addr: SupervisorPageAddr) -> MttL1EntryType {
+    let l2_entry = get_l2_entry(phys_addr.bits());
+    match l2_entry {
+        MttL1Dir(pfn) => {
+            let (l1_index, sub_index) = get_l1_index_and_subindex(phys_addr.bits());
+            let value = get_l1_value(pfn, l1_index);
+            MttL1EntryType::from_raw((value >> sub_index) & 0x3)
+        }
+        _ => unreachable!(),
+    }
+}
+
+fn get_l1_index_and_subindex(phys_addr: u64) -> (u64, u64) {
+    // Bits 25:17 for the 9-bit index, then 3-bit left shift for the 8-byte entry
+    let l1_index = (phys_addr & 0x03fe_0000) >> 14;
+    // 5-bit sub-index (16:12) with 1-bit left shift for the 2-bits per entry
+    let sub_index = (phys_addr & 0x1_f000) >> 11;
+    (l1_index, sub_index)
+}
+
+fn get_l1_value(pfn: u64, l1_index: u64) -> u64 {
+    let mtt_l1_page = pfn << 12;
+    let l1_entry_addr = (mtt_l1_page + l1_index) as *const u64;
+    // Safety: The 4K L1 page is allocated by the TSM, and are are using a 12-bit index.
+    unsafe { *l1_entry_addr }
+}
+
+fn update_l1_value(pfn: u64, phys_addr: SupervisorPageAddr, memory_type: MttMemoryType) {
+    let (l1_index, sub_index) = get_l1_index_and_subindex(phys_addr.bits());
+    let value = get_l1_value(pfn, l1_index);
+    // Clear the original 2-bits and OR in the new value
+    let value = (value & !(3 << sub_index)) | ((memory_type as u64) << sub_index);
+    let mtt_l1_page = pfn << 12;
+    let l1_entry_addr = (mtt_l1_page + l1_index) as *mut u64;
+    // Safety: The 4K L1 page is allocated by the TSM, and are are using a 12-bit index.
+    unsafe {
+        *l1_entry_addr = value;
+    };
+}
+
+// The L1 is a 4KB page that's allocated by the TSM to manage the memory type (non-confidential or confidential) at a 4K-page granularity.
+// This is done by dividing the 64MB region containing the page into 16KB regions of 4K each, and the 2-bit encoding in the L1 page pointed
+// to by the L2 entry determines the type.
+// The 12-bit index into the 4K page is computed using the 14-bits [25:12] of the physical address, and then shifting right by 2
+fn set_l1_entry(phys_addr: SupervisorPageAddr, memory_type: MttMemoryType) {
+    let l2_entry = get_l2_entry(phys_addr.bits());
+    match l2_entry {
+        MttL1Dir(pfn) => update_l1_value(pfn, phys_addr, memory_type),
+        _ => unreachable!(),
+    }
+}
+
+/// Changes the L2 type for 16 consecutive entries (each representing a 64MB range) for the 1GB range
+/// containing the address to the specified type.
+/// This can be called to consolidate previously fractured ranges.
+pub fn set_1gb_range_type(phys_addr: SupervisorPageAddr, memory_type: MttMemoryType) {
+    let range_start = PageSize::Size1G.round_up(phys_addr.bits());
```

Comment:
if the phys_addr passed is not a 1G aligned address, should this fn return an error?

Reply:
This depends on how we want to define the semantics of the API. The intention here was to define is such that given an arbitrary phys_addr in an 1GB range, the entire 1GB containing the address would be marked as that type (NC or C). We can equally change it to require the caller to round_up() to the 1GB boundary before passing it in (and then enforce the check either via the type system, or explicit check), but don't know if it's any better. 

This has been captured in the description:

```
/// Changes the L2 type for 16 consecutive entries (each representing a 64MB range) for the 1GB range
/// containing the address to the specified type.
/// This can be called to consolidate previously fractured ranges.
```